Add more policy drafts

drybjed opened the pull request.

drybjed committed:
Add more policy drafts

ypid commented:
Looking into it. I will open a pull request against https://github.com/drybjed/debops-policy/tree/add-docs

ganto commented on docs/code-signing.rst:
```diff
@@ -0,0 +1,10 @@
+
+The code used by the DebOps Project and commited in its repositories is
+expected to be signed by a valid GPG key of a DebOps Developer or DebOps
+Contributor. This should allow for secure code authentication and a valid audit
```
What does "is expected to be signed" mean? It must? Or it should? What if it isn't?

For my taste this is a bit exaggerated. As there is no controllable deployment mechanism (only git pull) or stable APIs (variable naming, sematics), I'm not really worried about someone bringing in unauthorized changes, but more about someone bringing in changes which will blow my setup. Therefore I need to do a manual code audit anyway before rolling out any changes to my environment. Signed commits won't save me from this. If it's optional, I'm fine with it.

I see that it could make sense for DebOps core developers which have full commit privileges on all repositories. But even there, if your laptop with your unprotected SSH private key and your GPG key is gone, even a commit signature won't protect the code. Or did I miss something? Maybe you should add a link somewhere why code signing makes sense in the first place. What happens if you revoke the key that was used for signing?

ypid replied:
I am currently working on this section. I guess you are right. It mainly affects the core devs and should be optional (but encouraged) for now for contributes as disused with @drybjed. To your questions how to use GPG and so one, there are good answers to those questions. To the OpSec of developers you mentioned. You are right. That is a difficult topic that can not be solved by GPG/PGP alone but it is a important part of it.

drybjed replied:
For key expiry, GPG keys can be stored in a "retired" keyring in the repository, similar to how [Debian does it](https://anonscm.debian.org/git/keyring/keyring.git/tree/). This could be used to authenticate commits signed by past keys.

Code signing for me is in part for a sane audit trail, so we can tell for example who merged a security hole, for example.

Other thing I'm thinking about is that since project is highly distributed (repositories on GitHub won't be there forever), moving the "trust" to GPG signatures stored with the `gitt` repository could allow for more secure distributed development. Some thoughts on code signing can be found in [A Git Horror Story](https://mikegerwitz.com/papers/git-horror-story) blog post. Since it's @ypid that convinced me to start signing my code, I imagine he could add his own thoughts here as well.

ypid replied:
On the other hand, creating a PGP key is not difficult. It should also be doable for contributes at least to a later time. The difficult part with PGP is to keep your private key private :wink:

Or as [Edward Snowden](https://www.wired.com/2014/10/snowdens-first-emails-to-poitras/) put it:

> Please confirm that no one has ever had a copy of your private key and that it uses a strong passphrase. Assume your adversary is capable of one trillion guesses per second.

drybjed replied:
@ypid Policy should contain some suggestions about key authentication. I imagine GPG keys signed by Developers should be fine, we could also rely on https://keybase.io/ and GitHub usage for Contributors.

ypid replied:
Sounds good. Here is my RFC: https://github.com/drybjed/debops-policy/pull/1/files

ypid replied:
> But even there, if your laptop with your unprotected SSH private key and your GPG key is gone, even a commit signature won't protect the code.

@ganto That is something I have been thinking about for which I will hopefully implement a solution someday in the future. To address this, I plan to swap out GPG subkeys more often and to destroy the private key of the GPG subkey’s after it has expired so that even I could not fake a signature after a certain time. Obviously I would need to rethink my backup strategy for my GPG keys for this to make sense.

Also, you are encouraged to protect your system as good as you can :wink:

ypid replied:
> more about someone bringing in changes which will blow my setup. Therefore I need to do a manual code audit anyway before rolling out any changes to my environment.

@ganto So I guess you are reviewing parts of DebOps already? Maybe it makes sense to take a similar approach as me who also does this by additionally signing the latest version of DebOps reviewed components using git submodules and GPG. This way, when someone trusts you, they can infer some of that trust onto the reviewed components. Or when multiple people do this who can somehow be trusted or are expected not to cooperate or the like that can also build up some trust or at least make it more difficult to introduce flaws and not get caught.

I recently documented this a bit more: https://github.com/ypid/ypid-ansible-common/

ypid committed:
Reworked code signing policy section.

Intended for: https://github.com/debops/debops-policy/pull/2

Incorporating the comments up until:
https://github.com/debops/debops-policy/pull/2#discussion_r67621711

ypid committed:
Fixed typo.

ypid committed:
Adjusted spelling to fit the rest of the documentation.

drybjed committed:
Merge branch 'ypid-add-docs/ypid-rework-gpg-signing' into add-docs

ypid committed:
Fixed spelling.

ypid committed:
Added access restriction and checking part to bots section.

ypid committed:
Added myself to the changelog.

ypid committed:
Added .gitignore file to allow easy docs builds.

ypid committed:
Dropped Role Contributors and moved responsibility to Role Maintainers.

drybjed committed:
Merge branch 'ypid-rework' into add-docs

ypid commented:
@drybjed How about adding this:

``` RST
Contributors
------------

Every person who is contributing to the DebOps Project.
The contributions need to be reviewed by one of the DebOps Developers.
```

drybjed committed:
Slight rewording, possibility of a Maintainer team

drybjed commented:
@ypid Sounds good.

ypid commented:
> I imagine GPG keys signed by Developers should be fine

@drybjed I guess you can add this too. I only mentioned Debian Developers.

drybjed committed:
Clarification about contributors, key signatures

ypid commented:
<!--
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512
- -->

@drybjed One thing about the current organizational structure. I am sure you are aware of the principle of least privilege and are already following it. Not sure how exactly this policy should be implemented but it sounds like you intent to give developers more access to the main repos? I think the current way of dealing with this is already pretty good e. g. you give developers only write access to places for example which they maintain or where they have contributed significant work. This also limits the impact of what developers can break. So for me currently, I would prefer to ask you when I think extra access would be useful and with proper reason. So basically, only give developers access if there is a good reason which you are already doing :+1:

BTW: Just an experiment. I signed this comment with GPG and hide the signature in HTML comments. I know you can edit this message so you can have a look. Maybe Git(Hub|Lab) can add some kind of support for this as well :) Maybe the full message could also be retrieved via the API and checked locally. Would be handy to integrate this in [github-backup](https://github.com/joeyh/github-backup).

<!--
-----BEGIN PGP SIGNATURE-----
iQIcBAEBCgAGBQJXabQRAAoJEIb9mAu/GkD45e0P/3eKFTSxcSeNDL0vdPgXpg7d
WEwDNCaqqco4zApAEP/BjQAgvvuClicaivVHvrNf0uVaJfSMpiR2CzdzzorQz04g
s7TxFepkVm0nt5XdT9KYofxebSiFr7kLSXaTfNAQsLWyBLnYe1Hgv2LaISoDM9ap
O57K6ccfBCmVa74fXPfixFcjYN+hlaCPWoqgV57YfBI4zKYRPxmZeglq88ocEMxx
FXeCTGydpbdVRJ/ey+n/0Jtq4ArQ+fpK9codZ3t8EWz5gh7z7Ln/u5T3Yyy5R25K
xRFeIv2t1CmzvdLdNVOZSFaJdKnvpIhmh2wiLu6f2B7RZH9bYYLlxHA/i0gM/PYL
xTOzsixCJQvcf5sLKf2rgMZRR3fcWq5mcs7pq0aqFbZM4nQNplxhLtjW8/o+AC+s
/F2cODKFBWLHn10k329yia1M/BHE+LH4IZlURokV0WY54gD78gJIYYvXCUmo8Wa3
NazgLuDscCsLxLzMGKeyh+5OfJFrF4ZqJoVJU2XRkQIZkHelSjbNlq/kz/2HtoX0
qErAY2kNEB3UfrrDsfnMzBP929wRkqQZZHFSjMkD3OaAURMqhdO5IPgc4KkXNW/R
TaOZSfE21AsOhCGO0+53D7xsdu8R8onG71eISSxzrczgPVuOeesgWsomso/3ZUbp
aIgPYFxmqPI2ItAICvxz
=n2CU
-----END PGP SIGNATURE-----
-->

drybjed commented:
Sure, current status quo is fine by me. When I get a role from someone else I give that person write access to that role, so that they still can work on it. The `debops-contrb` organization is meant to be a little more relaxed though, so that people can add their own roles there without me needing to intervene.

Very nice trick with the comment. I wonder how did you made it? Interesting bit for me is that I didn't have to remove the HTML comments for the signature to appear as valid - I thought that it would be invalid unless I remove that, so what gives?

Support for signature verification in GitHub comments would indeed be nice, although I imagine that having to write the comment in an external editor may dissuade some people.

ypid commented:
Sounds good. Thanks for the clarification.

> I thought that it would be invalid unless I remove that, so what gives?

As far as I know, GPG is only concerned about the text part in between the markers, example:

```
-----BEGIN PGP SIGNED MESSAGE-----
Text
-----BEGIN PGP SIGNATURE-----
```

Sure, usability is the problem. And also that the signatures are not verified locally but instead a centralized service would verify them. An local application would be needed for that.

drybjed commented:
Oh, I get it now. You included the closing/opening HTML comment tags in your signed message. Clever. :-)

ypid commented:
Thanks  :wink:

ypid commented:
@drybjed BTW. A more usable approach for signing comments which end up on GitHub is writing messages for PR in the git commit message. I am doing this for a while now and it is working great :wink: I basically write any longer text in Vim anyway.

Example:
- PR: https://github.com/debops/docs/pull/165
- Commit: https://github.com/debops/docs/pull/165/commits/833bd7d243cb2ca2e42ee41b93d747f1aa02a92e
